Push filters onto inner joins

rytaft opened the pull request:
This commit adds a simple pattern to push a filter onto an
inner join if it can't be pushed below. This is going to be
important for queries where the `ON` condition isn't explicitly
specified. For example, `SELECT * FROM a, b WHERE a.x = b.x`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/petermattis/opttoy/55)
<!-- Reviewable:end -->

RaduBerinde commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[v4/opt/factory.go, line 185 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L341z06q1X-DiGE9uES:-L341z07kGrSAf5sQ9C3:b77xofq) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/factory.go#L185)):*
> ```Go
> 
> func (f *Factory) appendListItem(list ListID, newItem GroupID) ListID {
> 	existingList := f.mem.lookupList(list)
> ```

I think lookupList should `return m.lists[id.offset : id.offset+id.len : id.offset + id.len]` and/or we should append to `existingList[:len(existingList):len(existingList)` or we may overwrite some other list that follows.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[v4/opt/norm/push_down.opt, line 33 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L3433FAebMqZW1bFBdQ:-L3433FAebMqZW1bFBdR:b5t84xq) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L33)):*
> ```
> # PushDownSelectJoin pushes a condition on to the join if it can't be pushed
> # below.
> [PushDownSelectJoin, Normalize]
> ```

I haven't thought this through fully, but is there some reason that we shouldn't *always* merge a Select operator with InnerJoin/InnerJoinApply, regardless of correlation?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[v4/opt/norm/push_down.opt, line 33 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L3433FAebMqZW1bFBdQ:-L345jt89zjVI7t_mOpI:b-e6o8lu) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L33)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I haven't thought this through fully, but is there some reason that we shouldn't *always* merge a Select operator with InnerJoin/InnerJoinApply, regardless of correlation?
</blockquote></details>

I don't think so. The current crdb code always merges any filters into the ON condition of an inner join.

---

*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L346-KAyPqSKUpkflVW:b-zcpizq) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
> ```
> [PushDownSelectJoin, Normalize]
> (Select
>     $input:(InnerJoin|InnerJoinApply $left:* $right:* (Filters $onList:*))
> ```

I don't have an actionable suggestion, but it's a bit unfortunate that the "on" condition shows up as "filters", in both the DSL and the format output (they're not the same except for inner joins).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L348idDU1HEYG8sCacT:b-a8x7gc) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't have an actionable suggestion, but it's a bit unfortunate that the "on" condition shows up as "filters", in both the DSL and the format output (they're not the same except for inner joins).
</blockquote></details>

I'm not sure I follow. Don't "on" conditions filter records? Also, aren't we going to change that to an AND list soon anyway?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L349KoOcRO6HUt5DJMv:b-lkrjpg) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'm not sure I follow. Don't "on" conditions filter records? Also, aren't we going to change that to an AND list soon anyway?
</blockquote></details>

In a way, but they filter records differently. E.g.
`SELECT a LEFT OUTER JOIN b ON a.x = b.x AND a.y > 1`
vs
 `SELECT a LEFT OUTER JOIN b ON a.x = b.x WHERE a.y > 1`
On other operators a "filters" child refers to a set of filter applied on the output of the operator (which is like the second case). But for joins it's the former case.

Yeah if we remove `filters` and just use an andOp this is moot. Though it would be useful to improve the format output to label each child (currently the "filters" operator name does that for better or for worse).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L34DtOO7trK758az2PU:br5cdt5) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

In a way, but they filter records differently. E.g.
`SELECT a LEFT OUTER JOIN b ON a.x = b.x AND a.y > 1`
vs
 `SELECT a LEFT OUTER JOIN b ON a.x = b.x WHERE a.y > 1`
On other operators a "filters" child refers to a set of filter applied on the output of the operator (which is like the second case). But for joins it's the former case.

Yeah if we remove `filters` and just use an andOp this is moot. Though it would be useful to improve the format output to label each child (currently the "filters" operator name does that for better or for worse).
</blockquote></details>

My assumption would be that we always add an AndOp at the top level of joins and selects, similar to how we always add a FiltersOp today (always, except in case of TrueOp or FalseOp). That makes it much easier to match, as you always know you've got an AndOp list to match against.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L34EoZnyJUJ7lDa1Yi9:bca3fhc) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

My assumption would be that we always add an AndOp at the top level of joins and selects, similar to how we always add a FiltersOp today (always, except in case of TrueOp or FalseOp). That makes it much easier to match, as you always know you've got an AndOp list to match against.
</blockquote></details>

Hm. There are downsides to that, e.g. we'll need special casing in a bunch of places to ignore a single-child andOp.

I see the problem of pattern matching. But do we really need awareness of a top-level AND in the patterns? This pattern could have `$input:(InnerJoin|InnerJoinApply $left:* $right:* $on:*)` and use a "merge-filters" function to combine filters (adding andOps when necessary).

In other cases we use patterns to treat conjunctions differently, but I argued elsewhere that splitting off filters could be a higher-level primitive so that we don't care about conjunctions in the patterns. Maybe I should try to tweak v4 to do something like this and see how it looks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L34HFyXdSI8SH-o_m7J:b5p1suj) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Hm. There are downsides to that, e.g. we'll need special casing in a bunch of places to ignore a single-child andOp.

I see the problem of pattern matching. But do we really need awareness of a top-level AND in the patterns? This pattern could have `$input:(InnerJoin|InnerJoinApply $left:* $right:* $on:*)` and use a "merge-filters" function to combine filters (adding andOps when necessary).

In other cases we use patterns to treat conjunctions differently, but I argued elsewhere that splitting off filters could be a higher-level primitive so that we don't care about conjunctions in the patterns. Maybe I should try to tweak v4 to do something like this and see how it looks.
</blockquote></details>

Yes, we'll have to try out different alternatives and see what works best. However, the list matching operator only works if the child actually has list children. For example:
```
(Select
    $input:*
    $filter:(And
        $list:[ ... $exists:(Exists $subquery:*) ... ]
    )
)
```
So if we don't always wrap filters with an And list, we'll have to duplicate this pattern to handle the "AND" case and the "not-AND" case (like single Exists expression). Possible, but might be kind of messy.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L34K6su7va8LUZtM10u:bhg6xdx) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Yes, we'll have to try out different alternatives and see what works best. However, the list matching operator only works if the child actually has list children. For example:
```
(Select
    $input:*
    $filter:(And
        $list:[ ... $exists:(Exists $subquery:*) ... ]
    )
)
```
So if we don't always wrap filters with an And list, we'll have to duplicate this pattern to handle the "AND" case and the "not-AND" case (like single Exists expression). Possible, but might be kind of messy.
</blockquote></details>

I see, yes, this is a case where we do need special behavior for a top-level AND (we can't hoist the subquery up as a join unless it's a top-level conjunction).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L34Lm2qniM2RXLFdacM:bffsia5) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I see, yes, this is a case where we do need special behavior for a top-level AND (we can't hoist the subquery up as a join unless it's a top-level conjunction).
</blockquote></details>

Hoisting subqueries outside of filter conditions is another case. I suspect we'll find a number of useful cases. v3 inlines the filters right into the exprLayout (and calls it filters, which is why I used that name). The trouble is that you *always* have to iterate over that filters list, even when it'd be more convenient to treat it as a single expression. That's why I liked FiltersOp. I can match it as a list, or treat it as a single expression, depending on the need. AndOp would give same advantage, as long as it was always present.

I'm also interested in your statement that we'd have to special case a bunch of places to ignore a single-child AndOp. Can you show me some examples where that would be necessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L34rnaDPnQQsMRU4ArH:bro5lbh) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Hoisting subqueries outside of filter conditions is another case. I suspect we'll find a number of useful cases. v3 inlines the filters right into the exprLayout (and calls it filters, which is why I used that name). The trouble is that you *always* have to iterate over that filters list, even when it'd be more convenient to treat it as a single expression. That's why I liked FiltersOp. I can match it as a list, or treat it as a single expression, depending on the need. AndOp would give same advantage, as long as it was always present.

I'm also interested in your statement that we'd have to special case a bunch of places to ignore a single-child AndOp. Can you show me some examples where that would be necessary?
</blockquote></details>

I was thinking of converting to a TypedExpr, though it seems pretty minor (in fact, it might not even need a special case).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks for the reviews!  

Sorry for the slow reply and sorry I missed all the discussion below - I need to fix my notifications for the opttoy repo :(

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[v4/opt/factory.go, line 185 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L341z06q1X-DiGE9uES:-L35A2ifitLZ1mPWqblC:bs3hiju) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/factory.go#L185)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think lookupList should `return m.lists[id.offset : id.offset+id.len : id.offset + id.len]` and/or we should append to `existingList[:len(existingList):len(existingList)` or we may overwrite some other list that follows.
</blockquote></details>

Not sure I understand this comment - should I make any change here, or is this suggestion for inside the `lookupList` function?

---

*[v4/opt/norm/push_down.opt, line 33 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L3433FAebMqZW1bFBdQ:-L359_NCuxZV_50JIsAx:b-ni6n9l) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L33)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't think so. The current crdb code always merges any filters into the ON condition of an inner join.
</blockquote></details>

I wrote it this way so that it wouldn't override the other push-down patterns.  But I guess since they are above this pattern in the file it shouldn't matter?

---

*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L358kg7Dvtq969p7TSu:bww7t8q) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I was thinking of converting to a TypedExpr, though it seems pretty minor (in fact, it might not even need a special case).
</blockquote></details>

For now, I suppose I should leave this pattern as-is with Filters?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55#-:-L35AExF5b1hOZR92w4k:b-qtx42k)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[v4/opt/factory.go, line 185 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L341z06q1X-DiGE9uES:-L35Gnu_dTZCy6IKWn7W:b-pmofi7) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/factory.go#L185)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Not sure I understand this comment - should I make any change here, or is this suggestion for inside the `lookupList` function?
</blockquote></details>

I believe that the code as written can overwrite other lists. We must make sure that append reallocates and does not use any spare capacity.

I proposed two possible fixes (one to lookupList, the other in this code). I would do the one in `lookupList` since it will be useful in other places.

---

*[v4/opt/norm/push_down.opt, line 35 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L346-KAyPqSKUpkflVV:-L35HGr-e-kq0mEvu2BI:bb77ufg) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L35)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

For now, I suppose I should leave this pattern as-is with Filters?
</blockquote></details>

Yes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->

rytaft commented:
I've made the suggested changes - ready for another look.

I've also fixed the issues with EXISTS caused by a rebase of the cockroach master branch.

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[v4/opt/norm/push_down.opt, line 33 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L3433FAebMqZW1bFBdQ:-L39cXWGBFJtGjiuwsZi:bo16hdf) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/norm/push_down.opt#L33)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I wrote it this way so that it wouldn't override the other push-down patterns.  But I guess since they are above this pattern in the file it shouldn't matter?
</blockquote></details>

I've changed the pattern so it pushes all filters onto the inner join and removes the select.  Now it doesn't need the AppendListItem function, so I removed that from `factory.go`.

---

*[v4/opt/factory.go, line 185 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L341z06q1X-DiGE9uES:-L39d3aNb0k-sgT3NcP4:b-oghnyn) ([raw file](https://github.com/petermattis/opttoy/blob/7e3ef32b1ca3cf284ed186850181d9e4a030a42b/v4/opt/factory.go#L185)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I believe that the code as written can overwrite other lists. We must make sure that append reallocates and does not use any spare capacity.

I proposed two possible fixes (one to lookupList, the other in this code). I would do the one in `lookupList` since it will be useful in other places.
</blockquote></details>

Ah ok thanks - sorry for being a bit dense.  I removed this function but I've made your suggested change to `lookupList`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55#-:-L39dL02ZrMEkE9VriC2:b-kgkf9a)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[v4/build/scope.go, line 366 at r2](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L39enrQjPXAwATJuLqR:-L39enrQjPXAwATJuLqS:b-nmbsum) ([raw file](https://github.com/petermattis/opttoy/blob/edc22ea4b328ab274a47ef049ec871aa4f81c61b/v4/build/scope.go#L366)):*
> ```Go
> 
> 	// is this an EXISTS statement?
> 	Exists bool
> ```

NIT: how come this is capitalized while other fields are not?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55#-:-L39fKLs8FjlYAXd13Xm:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[v4/build/scope.go, line 366 at r2](https://reviewable.io:443/reviews/petermattis/opttoy/55#-L39enrQjPXAwATJuLqR:-L39gUfzcYeFbgl0mG3f:b-7l2psr) ([raw file](https://github.com/petermattis/opttoy/blob/edc22ea4b328ab274a47ef049ec871aa4f81c61b/v4/build/scope.go#L366)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: how come this is capitalized while other fields are not?
</blockquote></details>

Good point - it doesn't need to be exported.  Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/55)*
<!-- Sent from Reviewable.io -->
